### PR TITLE
AWS: support overriding endpoint in DynamoDB

### PR DESCRIPTION
**Pull request by ebyhr**
This makes testing DynamoDB integration easily with docker (https://hub.docker.com/r/amazon/dynamodb-local/).

### CODE REVIEW

**Comment by amogh-jahagirdar** [comment]
Thanks for this! This change looks good to me

File: aws/src/integration/java/org/apache/iceberg/aws/TestDefaultAwsClientFactory.java [hunk 1]
```diff
@@ -54,4 +56,16 @@ public void testS3FileIoCredentialsOverride() {
...
+  @Test
+  public void testDynamoDbEndpointOverride() {
+    Map<String, String> properties = Maps.newHashMap();
+    properties.put(AwsProperties.DYNAMODB_ENDPOINT, "https://unknown:1234");
+    AwsClientFactory factory = AwsClientFactories.from(properties);
+    DynamoDbClient dynamoDbClient = factory.dynamo();
+    AssertHelpers.assertThrowsCause("Should refuse connection to unknown endpoint",
+            SdkClientException.class,
+            "Unable to execute HTTP request: unknown",
+            dynamoDbClient::listTables);
```
Comment (amogh-jahagirdar) [inline aws/src/integration/java/org/apache/iceberg/aws/TestDefaultAwsClientFactory.java:69]: Indentation

File: aws/src/main/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactory.java [hunk 2]
```diff
@@ -69,7 +70,10 @@ public KmsClient kms() {
 
   @Override
   public DynamoDbClient dynamo() {
-    return DynamoDbClient.builder().applyMutation(this::configure).build();
+    return DynamoDbClient.builder()
+            .applyMutation(this::configure)
+            .applyMutation(builder -> AwsClientFactories.configureEndpoint(builder, dynamoDbEndpoint))
+            .build();
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/AssumeRoleAwsClientFactory.java:76]: Indentation

File: aws/src/main/java/org/apache/iceberg/aws/AwsClientFactories.java [hunk 3]
```diff
@@ -111,7 +112,10 @@ public KmsClient kms() {
 
     @Override
     public DynamoDbClient dynamo() {
-      return DynamoDbClient.builder().httpClientBuilder(configureHttpClientBuilder(httpClientType)).build();
+      return DynamoDbClient.builder()
+              .httpClientBuilder(configureHttpClientBuilder(httpClientType))
+              .applyMutation(builder -> configureEndpoint(builder, dynamoDbEndpoint))
+              .build();
```
Comment (amogh-jahagirdar) [inline aws/src/main/java/org/apache/iceberg/aws/AwsClientFactories.java:118]: Indentation

**Review by amogh-jahagirdar — CHANGES_REQUESTED** [review]
Sorry about that, I was doing a second pass of review, and realized that indentation in some places were off, can we double check that?

**Comment by ebyhr** [comment]
@amogh-jahagirdar Could you run workflow if you have the permission?

**Review by jackye1995 — COMMENTED** [review]
looks good to me! enabling CI

**Comment by ebyhr** [comment]
@jackye1995 Gentle reminder.

**Comment by rdblue** [comment]
Thanks, @ebyhr! Sorry for the delay.

**Pull request merged**